urlb57eb64056_jacadenac_ISI_Othello_UN_tgz-pJ8-othello_OthelloJ8/src/logica/ia/algoritmosGeneticos/Estrategia.java

```diff
@@ -0,0 +1,435 @@
+				for(int j=0 ; j<genes.length ; j++){
+					try {
+						genes[j] = new IntegerGene(conf, 0, RANGO_GEN-pesoGenAnterior);
+						pesoGenAnterior=(int)Math.random()*(RANGO_GEN-pesoGenAnterior);
+					} catch (InvalidConfigurationException e) {
+						System.out.println("Error al crear los Genes en GeneticOperator");
+					}
+					AleloActual = (Integer)cromosoma.getGenes()[j].getAllele();
+	
+					//Si el cromosoma más apto se encuentra repetido, realiza mutaciones pequeńas sobre ellos
+					//Si no, realizará mutaciones grandes a unos y cruzes con otros
+					if(cromosoma.equals(cromosomaMasApto)){
+						/**///System.out.println("Entró a mutar en primer grado!");
+						valorAleatorio = (int) Math.floor(Math.random()*(10-130+1)+130); // Valor entre 0 y 80, ambos incluidos.
+						if (AleloActual  + (signo*valorAleatorio) < 0 || AleloActual  + (signo*valorAleatorio) > RANGO_GEN) { 
+							AleloNuevo = AleloActual +(-1)*(signo*valorAleatorio);
+						} else {
+							AleloNuevo = AleloActual +(-1)*(signo*valorAleatorio);
+						}
+					} else if( i%2 == 0 ){
+						/**///System.out.println("Entró a mutar en segundo grado!");
+						valorAleatorio = (int) Math.floor(Math.random()*(100-600+1)+600); //Valor entre 100 y 600, ambos incluidos.
+						if (AleloActual  + (signo*valorAleatorio) < 0 || AleloActual  + (signo*valorAleatorio) > RANGO_GEN) { 
+							AleloNuevo = AleloActual +(-1)*(signo*valorAleatorio);
+						} else {
+							AleloNuevo = AleloActual +(-1)*(signo*valorAleatorio);
+						}
+					} else {
+						if( j%2 == 0 && (i+1) < cromosomasActuales.size() && j<cromosomasActuales.get(i+1).getGenes().length){
+							/**///System.out.println("Entró a cruzar!");
+							int AleloSiguiente = (Integer)cromosomasActuales.get(i+1).getGenes()[j+1].getAllele();
+							AleloNuevo = AleloSiguiente;
+						} else {
+							AleloNuevo = AleloActual;
+						}
+					}
+					genes[j].setAllele(AleloNuevo);
+				}		
+				try {
+					IChromosome cromosomaNuevo = new Chromosome(conf, genes);
+					temporal.add(cromosomaNuevo);
+					//System.out.println("size="+temporal.size());
+				} catch (InvalidConfigurationException e) {
+					System.out.println("Error al ańadir a Cromosoma temporal en GeneticOperator");
+				}
+				i++;
+			}
+			int cantPoblacionAMutar = (int)(conf.getPopulationSize()/**0.9*/);
+			nueva.add(cromosomaMasApto);
+			for(int j=1; j < cantPoblacionAMutar; j++){
+				valorAleatorio = (int) Math.random()*temporal.size();
+				if(temporal !=null){
+					nueva.add(temporal.get(valorAleatorio));
+					temporal.remove(valorAleatorio);
+				}
+			}
+			return nueva;
+	}
+
+	private static void leerParametros() {
+		JFileChooser fileChooser = new JFileChooser();
+		String nombreFichero = "parametrosIniciales.txt";
+		File fichero = new File(directorio.getPath(),nombreFichero);
+		fileChooser.setCurrentDirectory(fichero);		
+		try {
+			BufferedReader reader = new BufferedReader(new FileReader(fichero));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *reader*. The resource is referenced at lines 422, 423. The resource is closed at line 427. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *reader* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).